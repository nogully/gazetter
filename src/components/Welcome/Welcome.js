import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { array, object, func } from 'prop-types';
import { auth, provider } from '../../firebase';
import { logIn, populateTweets } from '../../actions/actions';
import * as api from '../../apiCalls';


export class Welcome extends Component {
  constructor() {
    super();
    this.state = {
      loading: false
    };
  }

  signIn = async () =>  {  
    this.setState( {loading: true} );
    const user = await auth.signInWithPopup(provider)
    this.props.logIn(await user);
    const { credential } = await user;
    this.fetchTweets(credential)
    console.log(user)
  }

  fetchTweets = async (credential) => {
    const { accessToken, secret } = credential;
    try {
      const tweets = await api.getTweets();
      this.props.populateTweets(tweets);
      this.props.history.push('/news');
    } catch (error) {
      throw (error);
    }
  }

  render() {
    return (
      <div className="Welcome">
        <p id='tagline'>&quot;All the news that&apos;s fit to tweet&quot;</p>
        <button onClick={this.signIn}>Sign in with <i className="fab fa-twitter"></i> Twitter</button>
        { this.state.loading && !this.props.tweets.length ? 
          <img className="loading-gif"src="./newspapers.gif" alt="newspapers" /> : null }
      </div>
    );
  }
}

export const mapStateToProps = (store) => ({
  user: store.user,
  tweets: store.tweets
});

export const mapDispatchToProps = (dispatch) => ({
  logIn: user => dispatch(logIn(user)), 
  populateTweets: tweets => dispatch(populateTweets(tweets))
});

Welcome.propTypes = {
  tweets: array.isRequired, 
  user: object.isRequired,
  logIn: func.isRequired,
  populateTweets: func.isRequired, 
  history: object
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Welcome));