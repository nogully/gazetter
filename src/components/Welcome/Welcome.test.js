/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { Welcome, mapStateToProps, mapDispatchToProps } from './Welcome';
import { auth, provider } from '../../firebase';


describe('App', () => {
  it('should match snapshot', () => { 
    const wrapper = shallow(<Welcome tweets={ [] } user={ {} } logIn={ jest.fn() } populateTweets={ jest.fn() }/>);
    expect(wrapper).toMatchSnapshot();
  })

  describe('signIn', () => {
    let wrapper;

    beforeEach(() => {
      let mockTweets = [
              { id: 1234,
                full_text: 'Jhun named best teacher ever', 
                entities: { urls: [] }, 
                user: { screen_name: 'nytimes' }, 
                retweeted_status: 345,
                favorite_count: 456 }, 
              { id: 4567, 
                full_text: 'What\'s good, fam?', 
                entities: { urls: [] }, 
                user: { screen_name: 'yung-jhun' }, 
                retweeted_status: 0,
                favorite_count: 2  }
      ]
      window.fetch = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          json: () =>
            Promise.resolve(mockTweets)
        })
      })
      let mockUser = { display_name: 'noragully' }
    })

    it('component should start with a default state of loading: false', () => {
      wrapper = shallow(<Welcome tweets={ [] } user={ {} } logIn={ jest.fn() }
                                 populateTweets={ jest.fn() } />);
      expect(wrapper.state()).toEqual({loading: false});
    })

    it('should set state to loading: true', () => {
      wrapper = shallow(<Welcome tweets={ [] } user={ {} } logIn={ jest.fn() } 
                                 populateTweets={ jest.fn() } 
                                 history={[]}/>);
      expect(wrapper.state()).toEqual({loading: false})
      wrapper.instance().signIn();
      wrapper.update();
      expect(wrapper.state()).toEqual({loading: true})
    })

    it('should authenticate with Firebase, which returns a promised user', () => {
      wrapper = shallow(<Welcome tweets={ [] } user={ {} } logIn={ jest.fn() } 
                                 populateTweets={ jest.fn() } 
                                 history={[]}/>);
      auth.signInWithPopup = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          json: () =>
            Promise.resolve(mockUser)
        })
      })
      wrapper.instance().signIn();
      expect(auth.signInWithPopup).toHaveBeenCalledWith(provider);
    })
  })

  describe('fetchTweets', () => {
    it('should call an api', () => {
      
    })
  })

  describe('mapStateToProps', () => {
    it('should define the user props for the App container via Redux', () => {
      const mockUser = { display_name: 'noragully' }
      const mockStore = { user: mockUser }
      const expectedProps = { user: { display_name: 'noragully' } }
      const mapped = mapStateToProps(mockStore)
      expect(mapped).toEqual(expectedProps)
    })

    it('should define the tweets props for the App container via Redux', () => {
      const mockTweets = [
              { text: 'Why tho, Jhun?', 
                user: { display_name: 'noragully' } }, 
              { text: 'What\'s good, fam?', 
                user: { display_name: 'yung-jhun' } }
            ]
      const mockStore = { tweets: mockTweets };
      const expectedProps = { tweets: [
              { text: 'Why tho, Jhun?', 
                user: { display_name: 'noragully' } }, 
              { text: 'What\'s good, fam?', 
                user: { display_name: 'yung-jhun' } }
            ] }
      const mapped = mapStateToProps(mockStore);
      expect(mapped).toEqual(expectedProps)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call the Redux dispatch method with logIn', () => {
      const mockDispatch = jest.fn()
      const mapped = mapDispatchToProps(mockDispatch);
      mapped.logIn();
      expect(mockDispatch).toHaveBeenCalled();
    })

    it('should call the Redux dispatch method with populateTweets', () => {
      const mockDispatch = jest.fn()
      const mapped = mapDispatchToProps(mockDispatch);
      mapped.populateTweets();
      expect(mockDispatch).toHaveBeenCalled();
    })
  })
})