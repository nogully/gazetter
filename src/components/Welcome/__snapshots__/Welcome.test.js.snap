// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Welcome
    logIn={[Function]}
    populateTweets={[Function]}
    tweets={Array []}
    user={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          id="tagline"
>
          "All the news that's fit to tweet"
</p>,
        <button
          onClick={[Function]}
>
          Sign in with 
          <i
                    className="fab fa-twitter"
          />
           Twitter
</button>,
        null,
      ],
      "className": "Welcome",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "\\"All the news that's fit to tweet\\"",
          "id": "tagline",
        },
        "ref": null,
        "rendered": "\\"All the news that's fit to tweet\\"",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Sign in with ",
            <i
              className="fab fa-twitter"
/>,
            " Twitter",
          ],
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          "Sign in with ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fab fa-twitter",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          " Twitter",
        ],
        "type": "button",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            id="tagline"
>
            "All the news that's fit to tweet"
</p>,
          <button
            onClick={[Function]}
>
            Sign in with 
            <i
                        className="fab fa-twitter"
            />
             Twitter
</button>,
          null,
        ],
        "className": "Welcome",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "\\"All the news that's fit to tweet\\"",
            "id": "tagline",
          },
          "ref": null,
          "rendered": "\\"All the news that's fit to tweet\\"",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Sign in with ",
              <i
                className="fab fa-twitter"
/>,
              " Twitter",
            ],
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            "Sign in with ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fab fa-twitter",
              },
              "ref": null,
              "rendered": null,
              "type": "i",
            },
            " Twitter",
          ],
          "type": "button",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
